go/vt/discovery/topology_watcher_test.go

```diff
@@ -248,94 +222,77 @@ func checkWatcher(t *testing.T, refreshKnownTablets bool) {
 		key2 := TabletToMapKey(tablet2)
-		if _, ok := allTablets[key2]; !ok {
-			t.Fatalf("tablet was lost because it's reusing an address recently used by another tablet: %v", key2)
-		}
+		assert.Contains(t, allTablets, key2, "tablet was lost because it's reusing an address recently used by another tablet: %v", key2)
 
 		// Change tablets back to avoid altering later tests.
-		if _, err := ts.UpdateTabletFields(context.Background(), tablet2.Alias, func(t *topodatapb.Tablet) error {
+		_, err = ts.UpdateTabletFields(context.Background(), tablet2.Alias, func(t *topodatapb.Tablet) error {
 			t.Hostname = origTablet2.Hostname
 			t.PortMap = origTablet2.PortMap
 			tablet2 = t
 			return nil
-		}); err != nil {
-			t.Fatalf("UpdateTabletFields failed: %v", err)
-		}
-		if _, err := ts.UpdateTabletFields(context.Background(), tablet.Alias, func(t *topodatapb.Tablet) error {
+		})
+		require.Nil(t, err, "UpdateTabletFields failed")
+
+		_, err = ts.UpdateTabletFields(context.Background(), tablet.Alias, func(t *topodatapb.Tablet) error {
 			t.Hostname = origTablet.Hostname
 			tablet = t
 			return nil
-		}); err != nil {
-			t.Fatalf("UpdateTabletFields failed: %v", err)
-		}
+		})
+		require.Nil(t, err, "UpdateTabletFields failed")
+
 		tw.loadTablets()
-		counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "GetTablet": 2, "ReplaceTablet": 2})
+		counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "ReplaceTablet": 2})
 	}
 
 	// Remove the tablet and check that it is detected as being gone.
-	if err := ts.DeleteTablet(context.Background(), tablet.Alias); err != nil {
-		t.Fatalf("DeleteTablet failed: %v", err)
-	}
-	if _, err := topo.FixShardReplication(context.Background(), ts, logger, "aa", "keyspace", "shard"); err != nil {
-		t.Fatalf("FixShardReplication failed: %v", err)
-	}
+	require.NoError(t, ts.DeleteTablet(context.Background(), tablet.Alias))
+
+	_, err = topo.FixShardReplication(context.Background(), ts, logger, "aa", "keyspace", "shard")
+	require.Nil(t, err, "FixShardReplication failed")
 	tw.loadTablets()
-	if refreshKnownTablets {
-		counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "GetTablet": 1, "RemoveTablet": 1})
-	} else {
-		counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "RemoveTablet": 1})
-	}
+	counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "RemoveTablet": 1})
 	checkChecksum(t, tw, 789108290)
 
 	allTablets = fhc.GetAllTablets()
+	assert.Len(t, allTablets, 1)
 	key = TabletToMapKey(tablet)
-	if _, ok := allTablets[key]; ok || len(allTablets) != 1 {
-		t.Errorf("fhc.GetAllTablets() = %+v; don't want %v", allTablets, key)
-	}
+	assert.NotContains(t, allTablets, key)
+
 	key = TabletToMapKey(tablet2)
-	if _, ok := allTablets[key]; !ok || len(allTablets) != 1 || !proto.Equal(allTablets[key], tablet2) {
-		t.Errorf("fhc.GetAllTablets() = %+v; want %+v", allTablets, tablet2)
-	}
+	assert.Contains(t, allTablets, key)
```

Comment:
Curious why we use require in some places and assert in others? It's fine, I'm just curious. We want the test to fail immediately on the requires (instead of t.fatal), and we only want to log the failure and cause the eventual test failure (but continue on, instead of t.Errorf) when we use assert?

Reply:
That is exactly right. And it is intentional. I carefully went through all of them to make sure we are doing it right.
If we are unable to `CreateTablet` that's fatal. But if we are checking conditions on a result, there's no reason to not keep going.